loopy/tools.py

```diff
@@ -128,6 +128,16 @@ def update_for_pymbolic_expression(self, key_hash, key):
 
     update_for_PMap = update_for_dict  # noqa: N815
 
+    def update_for_OrderedDict(self, key_hash, key):  # noqa: N815
+        # Firedrake accidentally uses these in kernels. It shouldn't (and it
+        # doesn't notice because they disable caching), but we'll tolerate it
+        # for now.
```

Comment:
Where do they come from? We should probably fix on our end?

Reply:
Ah, probably here https://github.com/OP2/PyOP2/blob/3a1b62fe38a89ddee4e23bebf664a295f5538990/pyop2/codegen/rep2loopy.py#L535 (`parameters.temporaries` is an `OrderedDict`). If the dict is _not_ ordered, what guarantee (if any) does loopy provide on the order in which temporaries are declared in the resulting generated kernel?

Reply:
> If the dict is not ordered, what guarantee (if any) does loopy provide on the order in which temporaries are declared in the resulting generated kernel?

Deterministic, but unspecified. What guarantees would you like?

Reply:
deterministic across ranks I think (possibly it is no longer relevant because the on-disk key is not a function of the C-string generation @connorjward?)

Reply:
The cache key for the `.so` still depends on the C string. The disk caching I’ve recently added in Firedrake is for the outputs of TSFC so is unrelated to this.

I think it’s straightforward to replace the `OrderedDict` in PyOP2 with something loopy is happy with. A while ago I think I did just that to see whether enabling loopy caching gave us anything.

Reply:
> deterministic across ranks

Is that distinct from "just plain deterministic" (given a set of temp var names)? Doesn't sound like it, but want to make sure.

Reply:
> Is that distinct from "just plain deterministic" (given a set of temp var names)? Doesn't sound like it, but want to make sure.

No. As long as given `set(temporaries1) == set(temporaries2)` then `loopy_kernel(with_temporaries1)` and `loopy_kernel(with_temporaries2)` generate the same C string then all is good. (without relying on iteration order of python sets)

In the past we've collected various things using dictionaries and then emitted them in that order which has been bad (so lots of the code is littered with `OrderedDict`). I suppose if you sort the names that is sufficient.

Reply:
> No. As long as given `set(temporaries1) == set(temporaries2)` then `loopy_kernel(with_temporaries1)` and `loopy_kernel(with_temporaries2)` generate the same C string then all is good. (without relying on iteration order of python sets)

Great. That's precisely what I'd like to promise. If you see that violated somewhere, report it, it's a bug.

Reply:
@connorjward can you make the pyop2 change?

Reply:
Sure. I’ll have a go tomorrow.

Reply:
I think I've done it: https://github.com/OP2/PyOP2/pull/673